### ISSUE TITLE
FR: fail-closed setup guide

### ISSUE THREAD

**Issue by wizzard0** [tailscale/tailscale#4073]
### What are you trying to do?

Configure a PC so that it uses Tailscale exit node for apps other than Tailscale itself, and blocks all outgoing traffic otherwise.

Should definitely be doable but existing guides for OpenVPN etc are not directly applicable to Tailscale because e.g. tailscaled connects to many IPs, DERP nodes etc.

### How should we solve this?

Tailscale has outstanding documentation for other use cases, would love to have a guide for this use case too.

### What is the impact of not solving this?

Right now in this scenario one must resort to using OpenVPN on host machine for privacy, and restrict Tailscale for usage inside VM. Which results in added latency, resource use and generally defeats 2/3 of Tailscale selling points.

**Comment by DentonGentry** [comment]
I don't understand what more there would be in such a guide beyond "select the exit node to use." The default behavior when using an exit node is for all Internet traffic to go via the exit node and for LAN access to be blocked.

**Comment by wizzard0** [comment]
The desired behavior is to go completely offline when Tailscale is disconnected/connecting/reconnecting/daemon is starting/daemon crashed

**Comment by bradfitz** [comment]
On Windows I suppose we could have some `tailscale` CLI subcommand that installs the Windows Firewall rules that we use for exit nodes but in a mode that's not tied to the process lifetime.

/cc @danderson

**Comment by wizzard0** [comment]
@bradfitz yep CLI is fine, and it's completely fine to prototype as a separate script -- while I believe you can do it as seamless as the rest of Tailscale :) it's still a relatively power-user tool

**Comment by wizzard0** [comment]
Updated the issue with a list of test cases

**Comment by danderson** [comment]
The main difficulty, on all platforms, is to ensure that the blocking is fully accurate. This is much higher stakes than we normally do, so I'm hesitant to say we can do it (unlike other VPNs which happily do best-effort and then promise you that it's perfect).

On Windows, the most robust way of doing this would be to create "static" WFP filtering rules. Static rules persist after the lifetime of the process, and (iirc) are also reapplied on boot by Windows during network startup. This is in contrast to the "dynamic" rules we currently use, which are automatically cleaned up when the Tailscale service shuts down.

One complexity: the rules to allow VPN traffic have to reference the LUID of the network interface to allow, and that LUID doesn't exist until Tailscale starts and creates the interface. So, the "block everything" part is easy to implement, but the "except tailscale" part requires coordination with Tailscale in some way.

One final complication: the WFP framework is quite complex, and unless we find an expert to review our work, I'm not comfortable promising that absolutely nothing will leak out of the block (especially because various core Windows things install their own allow rules for various traffic - it's possible to override them, but it requires a bunch of care, especially if other e.g. A/V software is also trying to manage WFP rules.

I'll see if I can throw together something standalone to do this, but again I have to emphasize that "fail-closed" setups inside the OS are hard to get right. The most robust way to ensure full protection is to have two layers, one behind the other, and implement strict firewalling "outside" of the OS that you want to fail-closed. But obviously this is more complex and right now Tailscale can't help with that setup.

**Comment by danderson** [comment]
Did a bit of prototyping, but while the "block everything" part is relatively easy, "allow tailscale" is hard when running outside of Tailscale. You need to know Tailscale's interface LUID (which doesn't exist until tailscale is running), as well as the application ID of the Tailscale service, to allow it to use the underlay connections.

So, I don't think we can build this outside of Tailscale itself, annoyingly. And teaching Tailscale how to handle persistent firewall rules will take a bit more effort, so we'll need to figure out exactly what it takes.

**Comment by wizzard0** [comment]
@danderson hmm, interesting. Let me clarify: you see the main problem with allowing user apps (i.e. a web browser) to go thru tailscale when everything is blocked, or the Tailscale to connect?

> as well as the application ID of the Tailscale service, to allow it to use the underlay connections.

IIRC at least on Windows you can create firewall rules that filter by the tailscale(d) executable path? Don't know about Mac and pf though.
